plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/DevWatchContext.java

```diff
@@ -0,0 +1,522 @@
+        artifactFile = projectBuildDir.resolve(finalName + "." + ctx.getPackaging());
+        if (!Files.exists(ctx.getDeploymentsDir())) {
+            Files.createDirectories(ctx.getDeploymentsDir());
+        }
+        String fName = artifactFile.getFileName().toString();
+        boolean webapp = false;
+        boolean jarapp = false;
+        if (ctx.getPackaging().equals(WAR) || fName.endsWith(WAR)) {
+            if (ctx.isContextRoot()) {
+                fName = "ROOT." + WAR;
+            }
+            webapp = true;
+        } else {
+            if (ctx.getPackaging().equals(JAR) || fName.endsWith(JAR)) {
+                jarapp = true;
+            } else {
+                throw new RuntimeException("Not supported packaging : " + ctx.getPackaging());
+            }
+        }
+        isWebApp = webapp;
+        isJarApp = jarapp;
+        fileName = fName;
+        targetDir = ctx.getDeploymentsDir().resolve(fileName);
+
+        registerDir(ctx.getBaseDir());
+
+        pom = ctx.getPomFile();
+
+        for (Path p : ctx.getResources()) {
+            if (!p.isAbsolute()) {
+                p = ctx.getBaseDir().resolve(p);
+            }
+            // We must add it even if it doesn't exist.
+            // That the way to know the resource files in case they are created later.
+            resourceDirectories.add(p);
+            ctx.debug("[WATCH] resources dir: " + p);
+            if (Files.exists(p)) {
+                registerDir(p);
+            }
+        }
+
+        for (String extra : ctx.getExtraServerContent()) {
+            Path p = Paths.get(extra);
+            if (!p.isAbsolute()) {
+                p = ctx.getBaseDir().resolve(p);
+            }
+            extraDirectories.add(p);
+            ctx.debug("[WATCH] extra-content dir: " + p);
+            registerDir(p);
+        }
+
+        for (CliSession session : ctx.getCliSessions()) {
+            for (String f : session.getScriptFiles()) {
+                Path p = Paths.get(f);
+                if (!p.isAbsolute()) {
+                    p = ctx.getBaseDir().resolve(p);
+                }
+                cliFiles.add(p);
+                ctx.debug("[WATCH] CLI script File: " + p);
+                watchedDirectories.put(p.getParent().register(watcher, ENTRY_MODIFY), p.getParent());
+            }
+            if (session.getPropertiesFile() != null) {
+                Path p = Paths.get(session.getPropertiesFile());
+                if (!p.isAbsolute()) {
+                    p = ctx.getBaseDir().resolve(p);
+                }
+                cliFiles.add(p);
+                ctx.debug("[WATCH] CLI properties File: " + p);
+                watchedDirectories.put(p.getParent().register(watcher, ENTRY_MODIFY), p.getParent());
+            }
+        }
+    }
+
+    Path getTargetDirectory() {
+        return targetDir;
+    }
+
+    private void fileDeleted(Path absolutePath, Set<Path> paths) {
+        WatchKey key = null;
```

Comment:
Should this be surrounded in something like:
```
if (Files.isDirectory(absolutePath)) {
    // do stuff
}
```

Reply:
That is a directory yes, but the directory has been deleted, Files.isDirectory would return false. I think we need to check in all cases.

Reply:
Ah right of course!